lib/jni/PrivacyGuardState.cpp

```diff
@@ -0,0 +1,58 @@
+//
+// Copyright (c) 2015-2020 Microsoft Corporation and Contributors.
+// SPDX-License-Identifier: Apache-2.0
+//
+#include "PrivacyGuardState.hpp"
+#include "JniConvertors.hpp"
+
+namespace MAT_NS_BEGIN
+{
+
+/*static*/ CommonDataContext PrivacyGuardState::GenerateCommonDataContextObject(JNIEnv* env,
+                                                  jstring domainName,
+                                                  jstring machineName,
+                                                  jstring userName,
+                                                  jstring userAlias,
+                                                  jobjectArray ipAddresses,
+                                                  jobjectArray languageIdentifiers,
+                                                  jobjectArray machineIds,
+                                                  jobjectArray outOfScopeIdentifiers)
+{
+    CommonDataContext cdc;
+    cdc.DomainName = JStringToStdString(env, domainName);
+    cdc.MachineName = JStringToStdString(env, machineName);
+    cdc.UserName = JStringToStdString(env, userName);
+    cdc.UserAlias = JStringToStdString(env, userAlias);
+    cdc.IpAddresses = ConvertJObjectArrayToStdStringVector(env, ipAddresses);
+    cdc.LanguageIdentifiers = ConvertJObjectArrayToStdStringVector(env, languageIdentifiers);
+    cdc.MachineIds = ConvertJObjectArrayToStdStringVector(env, machineIds);
+    cdc.OutOfScopeIdentifiers = ConvertJObjectArrayToStdStringVector(env, outOfScopeIdentifiers);
+    return cdc;
+}
+
+/*static*/ std::shared_ptr<PrivacyGuard> PrivacyGuardState::getPrivacyGuardInstance()
+{
+    std::lock_guard<std::mutex> lock(m_pgInstanceGuard);
+    return m_pgInstance;
+}
+
+/*static*/ void PrivacyGuardState::setPrivacyGuardInstance(std::shared_ptr<PrivacyGuard> spPrivacyGuard)
+{
+    std::lock_guard<std::mutex> lock(m_pgInstanceGuard);
+    m_pgInstance.swap(spPrivacyGuard);
+}
+
+/*static*/ void PrivacyGuardState::resetPrivacyGuardInstance()
+{
+    std::lock_guard<std::mutex> lock(m_pgInstanceGuard);
+    m_pgInstance.reset();
+}
+
+/*static*/ bool PrivacyGuardState::isPrivacyGuardInstanceInitialized()
+{
+    std::lock_guard<std::mutex> lock(m_pgInstanceGuard);
+    return m_pgInstance != nullptr;
```

Comment:
@larvacea @saurabh-sagrawal Any thoughts on why I keep getting an undefined reference error for both `m_pgInstanceGuard` and `m_pgInstance` here when building it? Here's the exact error:
```
Execution failed for task ':maesdk:externalNativeBuildDebug'.
> Build command failed.
  Error while executing process C:\Users\sidahiy\AppData\Local\Android\Sdk\cmake\3.10.2.4988404\bin\ninja.exe with arguments {-C D:\GitRepos\cpp_client_telemetry\lib\android_build\maesdk\.cxx\cmake\debug\armeabi-v7a maesdk}
  ninja: Entering directory `D:\GitRepos\cpp_client_telemetry\lib\android_build\maesdk\.cxx\cmake\debug\armeabi-v7a'
  [1/1] Linking CXX shared library D:\GitRepos\cpp_client_telemetry\lib\android_build\maesdk\build\intermediates\cmake\debug\obj\armeabi-v7a\libmaesdk.so
  FAILED: D:/GitRepos/cpp_client_telemetry/lib/android_build/maesdk/build/intermediates/cmake/debug/obj/armeabi-v7a/libmaesdk.so 
  cmd.exe /C "cd . && C:\Users\sidahiy\AppData\Local\Android\Sdk\ndk\21.1.6352462\toolchains\llvm\prebuilt\windows-x86_64\bin\clang++.exe --target=armv7-none-linux-androideabi19 --gcc-toolchain=C:/Users/sidahiy/AppData/Local/Android/Sdk/ndk/21.1.6352462/toolchains/llvm/prebuilt/windows-x86_64 --sysroot=C:/Users/sidahiy/AppData/Local/Android/Sdk/ndk/21.1.6352462/toolchains/llvm/prebuilt/windows-x86_64/sysroot -fPIC -g -DANDROID -fdata-sections -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -D_FORTIFY_SOURCE=2 -march=armv7-a -mthumb -Wformat -Werror=format-security   -O0 -fno-limit-debug-info  -Wl,--exclude-libs,libgcc.a -Wl,--exclude-libs,libgcc_real.a -Wl,--exclude-libs,libatomic.a -Wl,--build-id -Wl,--fatal-warnings -Wl,--exclude-libs,libunwind.a -Wl,--no-undefined -Qunused-arguments -shared -Wl,-soname,libmaesdk.so -o D:\GitRepos\cpp_client_telemetry\lib\android_build\maesdk\build\intermediates\cmake\debug\obj\armeabi-v7a\libmaesdk.so @CMakeFiles/maesdk.rsp  && cd ."
  D:/GitRepos/cpp_client_telemetry/lib/jni/PrivacyGuardState.cpp:0: error: undefined reference to 'Microsoft::Applications::Events::PrivacyGuardState::m_pgInstanceGuard'
  D:/GitRepos/cpp_client_telemetry/lib/jni/PrivacyGuardState.cpp:0: error: undefined reference to 'Microsoft::Applications::Events::PrivacyGuardState::m_pgInstance'
  D:/GitRepos/cpp_client_telemetry/lib/jni/PrivacyGuardState.cpp:0: error: undefined reference to 'Microsoft::Applications::Events::PrivacyGuardState::m_pgInstanceGuard'
  D:/GitRepos/cpp_client_telemetry/lib/jni/PrivacyGuardState.cpp:0: error: undefined reference to 'Microsoft::Applications::Events::PrivacyGuardState::m_pgInstance'
  D:/GitRepos/cpp_client_telemetry/lib/jni/PrivacyGuardState.cpp:0: error: undefined reference to 'Microsoft::Applications::Events::PrivacyGuardState::m_pgInstanceGuard'
  D:/GitRepos/cpp_client_telemetry/lib/jni/PrivacyGuardState.cpp:0: error: undefined reference to 'Microsoft::Applications::Events::PrivacyGuardState::m_pgInstance'
  D:/GitRepos/cpp_client_telemetry/lib/jni/PrivacyGuardState.cpp:0: error: undefined reference to 'Microsoft::Applications::Events::PrivacyGuardState::m_pgInstanceGuard'
  D:/GitRepos/cpp_client_telemetry/lib/jni/PrivacyGuardState.cpp:0: error: undefined reference to 'Microsoft::Applications::Events::PrivacyGuardState::m_pgInstance'
  clang++: error: linker command failed with exit code 1 (use -v to see invocation)
  ninja: build stopped: subcommand failed.
```

Reply:
Note that the checks don't build with modules which makes them all pass here.

Reply:
You're defining a class scoped static object, but never defining it.
There should be a:
`static MAT::PrivacyGuardState::m_pgInstanceGuard;` in the .cpp file.